File: palette.py
# palette.py
import sys
import argparse
import colorsys
import json
import re

def hex_to_rgb(hex_color):
    hex_color = hex_color.lstrip('#')
    if len(hex_color) == 3:
        hex_color = ''.join(c*2 for c in hex_color)
    return tuple(int(hex_color[i:i+2], 16) for i in (0, 2, 4))

def rgb_to_hex(rgb):
    return '#{:02x}{:02x}{:02x}'.format(*rgb)

def rgb_to_hsl(rgb):
    r, g, b = [x/255.0 for x in rgb]
    h, l, s = colorsys.rgb_to_hls(r, g, b)
    return (h, s, l)  # h, saturation, lightness

def hsl_to_rgb(h, s, l):
    r, g, b = colorsys.hls_to_rgb(h, l, s)
    return tuple(int(x*255) for x in (r, g, b))

def generate_palette(hex_color):
    rgb = hex_to_rgb(hex_color)
    h, s, l = rgb_to_hsl(rgb)

    # Оттенки 50-900
    lightness_levels = [0.90, 0.82, 0.74, 0.66, 0.58, 0.50, 0.42, 0.34, 0.26, 0.18]
    shades = {}
    for i, lvl in enumerate(lightness_levels):
        name = 50 + i*100
        r, g, b = hsl_to_rgb(h, s, lvl)
        shades[name] = rgb_to_hex((r, g, b))

    # Акцентные цвета
    accent_lightness = [0.60, 0.70, 0.75, 0.80]
    accent_saturation = [0.95, 0.90, 0.85, 0.80]
    accent_names = ['A100', 'A200', 'A400', 'A700']
    accents = {}
    for i, name in enumerate(accent_names):
        r, g, b = hsl_to_rgb(h, accent_saturation[i], accent_lightness[i])
        accents[name] = rgb_to_hex((r, g, b))

    # Контрастный текст для каждого оттенка (белый или чёрный)
    contrast = {}
    for name, hex_c in shades.items():
        r, g, b = hex_to_rgb(hex_c)
        luminance = (0.299*r + 0.587*g + 0.114*b) / 255
        contrast[name] = '#ffffff' if luminance < 0.5 else '#000000'

    return shades, accents, contrast

def print_palette(shades, accents, contrast):
    # Заголовок
    print("\n--- Material Palette ---")
    # Оттенки
    for name, hex_c in shades.items():
        r, g, b = hex_to_rgb(hex_c)
        # Цветной блок с текстом
        text_color = contrast[name]
        # Используем ANSI True Color для фона и текста
        bg = f"\033[48;2;{r};{g};{b}m"
        fg = f"\033[38;2;{int(text_color[1:3],16)};{int(text_color[3:5],16)};{int(text_color[5:7],16)}m"
        reset = "\033[0m"
        print(f"{bg}{fg} {name:4} {hex_c} {reset}")

    # Акценты
    print("\n--- Accents ---")
    for name, hex_c in accents.items():
        r, g, b = hex_to_rgb(hex_c)
        bg = f"\033[48;2;{r};{g};{b}m"
        print(f"{bg} {name:4} {hex_c} {reset}")

def main():
    parser = argparse.ArgumentParser(description='Material Palette Generator')
    parser.add_argument('color', help='Основной цвет (HEX, например #6200EE)')
    parser.add_argument('--json', action='store_true', help='Вывод в JSON')
    parser.add_argument('--css', action='store_true', help='Вывод CSS-переменных')
    parser.add_argument('--scss', action='store_true', help='Вывод SCSS-переменных')
    args = parser.parse_args()

    try:
        shades, accents, contrast = generate_palette(args.color)
    except Exception as e:
        print(f"Ошибка: {e}", file=sys.stderr)
        sys.exit(1)

    if args.json:
        palette = {'shades': shades, 'accents': accents, 'contrast': contrast}
        print(json.dumps(palette, indent=2))
    elif args.css:
        print(":root {")
        for name, hex_c in shades.items():
            print(f"  --md-{name}: {hex_c};")
        for name, hex_c in accents.items():
            print(f"  --md-{name}: {hex_c};")
        print("}")
    elif args.scss:
        for name, hex_c in shades.items():
            print(f"$md-{name}: {hex_c};")
        for name, hex_c in accents.items():
            print(f"$md-{name}: {hex_c};")
    else:
        print_palette(shades, accents, contrast)

if __name__ == '__main__':
    main()

File: palette.cpp
// palette.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <cstdlib>

using namespace std;

struct RGB { int r, g, b; };
struct HSL { double h, s, l; };

RGB hexToRGB(const string& hex) {
    string s = hex;
    if (s[0] == '#') s = s.substr(1);
    if (s.length() == 3) {
        string t = "";
        for (char c : s) t += c + c;
        s = t;
    }
    int r, g, b;
    stringstream ss;
    ss << hex << s.substr(0,2);
    ss >> r;
    ss.clear();
    ss << hex << s.substr(2,2);
    ss >> g;
    ss.clear();
    ss << hex << s.substr(4,2);
    ss >> b;
    return {r, g, b};
}

string rgbToHex(int r, int g, int b) {
    stringstream ss;
    ss << "#" << setfill('0') << setw(2) << hex << r
       << setw(2) << hex << g << setw(2) << hex << b;
    return ss.str();
}

HSL rgbToHSL(int r, int g, int b) {
    double rf = r/255.0, gf = g/255.0, bf = b/255.0;
    double max = std::max({rf, gf, bf});
    double min = std::min({rf, gf, bf});
    double l = (max + min) / 2.0;
    double h, s;
    if (max == min) {
        h = 0; s = 0;
    } else {
        double diff = max - min;
        s = (l < 0.5) ? diff / (max + min) : diff / (2.0 - max - min);
        if (max == rf) {
            h = (gf - bf) / diff + (gf < bf ? 6.0 : 0.0);
        } else if (max == gf) {
            h = (bf - rf) / diff + 2.0;
        } else {
            h = (rf - gf) / diff + 4.0;
        }
        h /= 6.0;
    }
    return {h, s, l};
}

RGB hslToRGB(double h, double s, double l) {
    double r, g, b;
    if (s == 0) {
        r = g = b = l;
    } else {
        double v2 = (l < 0.5) ? l * (1 + s) : l + s - l*s;
        double v1 = 2*l - v2;
        auto hueToRGB = [&](double hh) {
            if (hh < 0) hh += 1;
            if (hh > 1) hh -= 1;
            if (6*hh < 1) return v1 + (v2-v1)*6*hh;
            if (2*hh < 1) return v2;
            if (3*hh < 2) return v1 + (v2-v1)*(2.0/3.0 - hh)*6;
            return v1;
        };
        r = hueToRGB(h + 1.0/3.0);
        g = hueToRGB(h);
        b = hueToRGB(h - 1.0/3.0);
    }
    return {(int)(r*255), (int)(g*255), (int)(b*255)};
}

void generatePalette(const string& hex, map<string, string>& shades, map<string, string>& accents, map<string, string>& contrast) {
    RGB rgb = hexToRGB(hex);
    HSL hsl = rgbToHSL(rgb.r, rgb.g, rgb.b);

    vector<double> lightness = {0.90, 0.82, 0.74, 0.66, 0.58, 0.50, 0.42, 0.34, 0.26, 0.18};
    vector<int> names = {50, 100, 200, 300, 400, 500, 600, 700, 800, 900};
    for (size_t i=0; i<lightness.size(); ++i) {
        RGB c = hslToRGB(hsl.h, hsl.s, lightness[i]);
        shades[to_string(names[i])] = rgbToHex(c.r, c.g, c.b);
    }

    vector<double> accentL = {0.60, 0.70, 0.75, 0.80};
    vector<double> accentS = {0.95, 0.90, 0.85, 0.80};
    vector<string> accentNames = {"A100", "A200", "A400", "A700"};
    for (size_t i=0; i<accentNames.size(); ++i) {
        RGB c = hslToRGB(hsl.h, accentS[i], accentL[i]);
        accents[accentNames[i]] = rgbToHex(c.r, c.g, c.b);
    }

    for (auto& p : shades) {
        RGB c = hexToRGB(p.second);
        double lum = (0.299*c.r + 0.587*c.g + 0.114*c.b) / 255.0;
        contrast[p.first] = (lum < 0.5) ? "#ffffff" : "#000000";
    }
}

void printPalette(const map<string, string>& shades, const map<string, string>& accents, const map<string, string>& contrast) {
    cout << "\n--- Material Palette ---" << endl;
    for (auto& p : shades) {
        RGB bg = hexToRGB(p.second);
        RGB fg = hexToRGB(contrast.at(p.first));
        cout << "\033[48;2;" << bg.r << ";" << bg.g << ";" << bg.b << "m"
             << "\033[38;2;" << fg.r << ";" << fg.g << ";" << fg.b << "m"
             << " " << setw(4) << p.first << " " << p.second << " \033[0m" << endl;
    }
    cout << "\n--- Accents ---" << endl;
    for (auto& p : accents) {
        RGB bg = hexToRGB(p.second);
        cout << "\033[48;2;" << bg.r << ";" << bg.g << ";" << bg.b << "m"
             << " " << setw(4) << p.first << " " << p.second << " \033[0m" << endl;
    }
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "Использование: palette <HEX-цвет> [--json] [--css] [--scss]" << endl;
        return 1;
    }
    string color = argv[1];
    bool jsonFlag = false, cssFlag = false, scssFlag = false;
    for (int i=2; i<argc; ++i) {
        string arg = argv[i];
        if (arg == "--json") jsonFlag = true;
        else if (arg == "--css") cssFlag = true;
        else if (arg == "--scss") scssFlag = true;
    }

    map<string, string> shades, accents, contrast;
    try {
        generatePalette(color, shades, accents, contrast);
    } catch (...) {
        cerr << "Ошибка генерации палитры" << endl;
        return 1;
    }

    if (jsonFlag) {
        cout << "{\"shades\":{";
        int i=0;
        for (auto& p : shades) {
            if (i++) cout << ",";
            cout << "\"" << p.first << "\":\"" << p.second << "\"";
        }
        cout << "},\"accents\":{";
        i=0;
        for (auto& p : accents) {
            if (i++) cout << ",";
            cout << "\"" << p.first << "\":\"" << p.second << "\"";
        }
        cout << "},\"contrast\":{";
        i=0;
        for (auto& p : contrast) {
            if (i++) cout << ",";
            cout << "\"" << p.first << "\":\"" << p.second << "\"";
        }
        cout << "}}\n";
    } else if (cssFlag) {
        cout << ":root {" << endl;
        for (auto& p : shades) cout << "  --md-" << p.first << ": " << p.second << ";" << endl;
        for (auto& p : accents) cout << "  --md-" << p.first << ": " << p.second << ";" << endl;
        cout << "}" << endl;
    } else if (scssFlag) {
        for (auto& p : shades) cout << "$md-" << p.first << ": " << p.second << ";" << endl;
        for (auto& p : accents) cout << "$md-" << p.first << ": " << p.second << ";" << endl;
    } else {
        printPalette(shades, accents, contrast);
    }
    return 0;
}
